cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/MetricSampleAggregator.java

```diff
@@ -0,0 +1,474 @@
+   *                              cache.
+   * @param metricDef metric definitions.
+   */
+  public MetricSampleAggregator(int numWindows,
+                                long windowMs,
+                                int minSamplesPerWindow,
+                                int maxAllowedImputations,
+                                int completenessCacheSize,
+                                MetricDef metricDef) {
+    super(0);
+    _identityEntityMap = new ConcurrentHashMap<>();
+    _rawMetrics = new ConcurrentHashMap<>();
+    _numWindows = numWindows;
+    _windowMs = windowMs;
+    // We keep one more window for the active window.
+    _numWindowsToKeep = _numWindows + 1;
+    _minSamplesPerWindow = minSamplesPerWindow;
+    _windowRollingLock = new ReentrantLock();
+    _maxAllowedImputations = maxAllowedImputations;
+    _metricDef = metricDef;
+    _aggregatorState = new MetricSampleAggregatorState<>(_generation.get(), _windowMs, completenessCacheSize);
+    _oldestWindowIndex = 0L;
+    _currentWindowIndex = 0L;
+  }
+
+  /**
+   * Add a sample to the metric aggregator.
+   *
+   * @param sample The metric sample to add.
+   *
+   * @return true if the sample is accepted, false if the sample is ignored.
+   */
+  public boolean addSample(MetricSample<G, E> sample) {
+    if (!sample.isValid(_metricDef)) {
+      LOG.warn("The metric sample is discarded due to missing metrics. Sample: {}", sample);
+      return false;
+    }
+    long windowIndex = windowIndex(sample.sampleTime());
+    // Skip the samples that are too old.
+    if (windowIndex < _oldestWindowIndex) {
+      return false;
+    }
+    boolean newWindowRolledOut = rollOutNewWindowIfNeeded(windowIndex);
+    RawMetricValues rawMetricValues =
+        _rawMetrics.computeIfAbsent(identity(sample.entity()), k -> {
+          RawMetricValues rawValues = new RawMetricValues(_numWindowsToKeep, _minSamplesPerWindow);
+          rawValues.updateOldestWindowIndex(_oldestWindowIndex);
+          return rawValues;
+        });
+    LOG.trace("Adding sample {} to window index {}", sample, windowIndex);
+    rawMetricValues.addSample(sample, windowIndex, _metricDef);
+    if (newWindowRolledOut || windowIndex != _currentWindowIndex) {
+      long generation = _generation.incrementAndGet();
+      // Data has been inserted to an old window.
+      if (windowIndex != _currentWindowIndex) {
+        _aggregatorState.updateWindowGeneration(windowIndex, generation);
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Aggregate the metric samples in the given period into a {@link MetricSampleAggregationResult} based on the
+   * specified {@link AggregationOptions}.
+   *
+   * <p>
+   *   The aggregation result contains all the entities in {@link AggregationOptions#interestedEntities()}.
+   *   For the entities that are completely missing, an empty result is added with all the value set to 0.0 and
```

Comment:
Initially it was because 0 is the entities that 0 has the least impact on the total result. Since the invalid entities are returned in the aggregation result, so the users should be able to know which entity is invalid.